packages/cli/src/lib/merge/index.js

```diff
@@ -0,0 +1,222 @@
+    // eslint-disable-next-line no-fallthrough
+    case 'Program':
+    case 'ArrayExpression':
+      return path.remove()
+  }
+}
+
+// When merging, trailing comments are a bit nasty. A comment can be parsed as a leading comment
+// of one expression, and a trailing comment of a subsequent expression. This is sort of an open
+// issue for Babel: https://github.com/babel/babel/issues/7002, but we can work around it pretty
+// easily with the following:
+function stripTrailingCommentsStrategy() {
+  return {
+    enter(path) {
+      path.node.trailingComments = []
+    },
+  }
+}
+
+/**
+ * 1. Traverse extAST and track the semantic IDs of all of the nodes for which we have a
+ *    merge strategy.
+ * 2. Traverse baseAST. On node exit, attempt to merge semantically-equivalent ext nodes.
+ *     a. When a semantically equivalent ext node is merged, it is pruned from ext.
+ * 3. Traverse extAST's body (if any nodes remain) and attempt to put top-level declarations
+ *    at their latest-possible positions.
+ *  a. Latest-possible is defined as the position immediately preceeding the first use of the
+ *     node's binding, if it exists.
+ */
+function mergeAST(baseAST, extAST, strategy = {}) {
+  const identity = extractProperty('identity', strategy) || semanticIdentity
+  const identities = {}
+  const baseVisitor = { ...stripTrailingCommentsStrategy() }
+  const extVisitor = { ...stripTrailingCommentsStrategy() }
+
+  forEachFunctionOn(strategy, (typename, strat) => {
+    extVisitor[typename] = {
+      enter(path) {
+        const id = identity(path)
+        id && (identities[id] ||= []).push(path)
+      },
+    }
+    baseVisitor[typename] = {
+      enter(path) {
+        if (isOpaque(strat)) {
+          skipChildren(path)
+        }
+      },
+      exit(path) {
+        const exts = extractProperty(identity(path), identities)
+        if (exts) {
+          const proxyPath = makeProxy(path)
+          exts.map(makeProxy).forEach((ext) => {
+            strat(proxyPath, ext)
+            prune(ext.path)
+          })
+        }
+      },
+    }
+  })
+
+  traverse(extAST, extVisitor)
+  traverse(baseAST, baseVisitor)
+
+  const baseProgram = getProgramPath(baseAST)
+  const [imports, others] = _.partition(
+    getProgramPath(extAST).get('body'),
+    nodeIs('ImportDeclaration')
+  )
+
+  imports.forEach((exp) => insertAfterLastImport(exp, baseProgram))
+  _.forEachRight(others, (exp) => insertBeforeFirstUsage(exp, baseProgram))
+}
+
+export function merge(base, extension, strategy) {
+  function parseReact(code) {
+    return parse(code, {
+      presets: ['@babel/preset-react'],
+    })
+  }
+
+  const baseAST = parseReact(base)
+  const extAST = parseReact(extension)
+
+  mergeAST(baseAST, extAST, strategy)
+  const { code } = generate(baseAST)
+
+  // When testing, use prettier here to produce predictable outputs.
+  // Otherwise, leave formatting to the caller.
+  return process.env.JEST_WORKER_ID
```

Comment:
Just realized I could move the formatting into the tests. Probably cleaner.

Reply:
Yeah, let's do that please :)